src/frontend/src/utils/condition.rs

```diff
@@ -239,6 +240,77 @@ impl Condition {
         .unwrap()
     }
 
+    /// generate range scans from each arm of `OR` clause and merge them
+    /// Currently, only support equal type range scans.
+    /// Keep in mind that range scans can not be overlap, otherwise duplicate rows will occur.
+    fn disjunctions_to_scan_ranges(
+        order_column_ids: &[usize],
+        num_cols: usize,
+        disjunctions: Vec<ExprImpl>,
+    ) -> Result<Option<(Vec<ScanRange>, Self)>> {
+        let disjunctions_result: Result<Vec<(Vec<ScanRange>, Self)>> = disjunctions
+            .into_iter()
+            .map(|x| {
+                Condition {
+                    conjunctions: to_conjunctions(x),
+                }
+                .split_to_scan_ranges(order_column_ids, num_cols)
+            })
+            .collect();
+
+        // if any arm of `OR` clause fail, bail out.
+        let disjunctions_result = disjunctions_result?;
+
+        // if all arm of `OR` clause scan ranges are simply equal condition type, merge all
+        // of them.
+        let all_equal = disjunctions_result
+            .iter()
+            .all(|(scan_ranges, other_condition)| {
+                other_condition.always_true()
+                    && scan_ranges
+                        .iter()
+                        .all(|x| !x.eq_conds.is_empty() && is_full_range(&x.range))
+            });
+
+        if all_equal {
+            // think about the case (a = 1) or (a = 1 and b = 2)
+            // we should only keep the large one range scan a = 1, because a = 1 is overlap with (a
+            // = 1 and b = 2)
+            let scan_ranges = disjunctions_result
+                .into_iter()
+                .flat_map(|(scan_ranges, _)| scan_ranges)
+                // sort, large one first
+                .sorted_by(|a, b| a.eq_conds.len().cmp(&b.eq_conds.len()))
+                .collect_vec();
+            // make sure each range is never overlap with others, that's what scan range mean
+            let mut non_overlap_scan_ranges: Vec<ScanRange> = vec![];
+            for s1 in &scan_ranges {
+                let mut overlap = false;
+                for s2 in &mut non_overlap_scan_ranges {
+                    let min_len = min(s1.eq_conds.len(), s2.eq_conds.len());
+                    overlap = s1
+                        .eq_conds
+                        .iter()
+                        .take(min_len)
+                        .zip_eq(s2.eq_conds.iter().take(min_len))
+                        .all(|(a, b)| a == b);
```

Comment:
Simply using `zip` here is OK? Can add `#[expect(clippy::disallowed_methods)]`

Reply:
I want to compare the prefix of s1 and s2 here. `zip` seems exactly what I want after I figure out its behavior.